test_http.py

```diff
@@ -0,0 +1,124 @@
+
+
+def test_parse_request_uri():
+    ''' Test parse request with a good request. '''
+    method = 'GET'
+    uri = '/index.html'
+    protocol = 'HTTP/1.1'
+    first_line = '{} {} {}'.format(method, uri, protocol)
+    timestamp = 'Date: ' + email.utils.formatdate(usegmt=True)
+    header = 'Host: www.example.com'
+    client_request = [first_line, timestamp, header, ' ', '\r\n']
+    assert es.parse_request('\r\n'.join(client_request)) == '/index.html'
+
+
+def test_client_socket_function_ok():
+    ''' Test a good request'''
+    method = 'GET'
+    uri = '/index.html'
+    protocol = 'HTTP/1.1'
+    first_line = '{} {} {}'.format(method, uri, protocol)
+    timestamp = 'Date: ' + email.utils.formatdate(usegmt=True)
+    header = 'Host: www.example.com'
+    client_request = [first_line, timestamp, header, ' ', '\r\n']
+
+    response = uri
+
+    recieve = ec.client_socket_function('\r\n'.join(client_request))
+    assert recieve == response
+
+
+def test_client_socket_function_405():
+    ''' Test a 405 error'''
+    method = 'PUSH'
+    uri = '/index.html'
+    protocol = 'HTTP/1.1'
+    first_line = '{} {} {}'.format(method, uri, protocol)
+    timestamp = 'Date: ' + email.utils.formatdate(usegmt=True)
+    header = 'Host: www.example.com'
+    client_request = [first_line, timestamp, header, ' ', '\r\n']
+
+    error_code = '405'
+    error_message = 'Method Not Allowed'
+    first_line = 'HTTP/1.1 {} {}'.format(error_code, error_message)
+    timestamp = 'Date: ' + email.utils.formatdate(usegmt=True)
+    content_header = 'Content-Type: text/plain'
+    body = '{} {}'.format(error_code, error_message)
+    response_list = [first_line, timestamp, content_header, ' ', body, '\r\n']
+
+    recieve = ec.client_socket_function('\r\n'.join(client_request))
+    assert recieve == '\r\n'.join(response_list)
+
+
+def test_client_socket_function_505():
+    ''' Test a 505 error'''
+    method = 'GET'
+    uri = '/index.html'
+    protocol = 'HTTP/1.0'
+    first_line = '{} {} {}'.format(method, uri, protocol)
+    timestamp = 'Date: ' + email.utils.formatdate(usegmt=True)
+    header = 'Host: www.example.com'
+    client_request = [first_line, timestamp, header, ' ', '\r\n']
+
+    error_code = '505'
+    error_message = 'HTTP Version Not Supported'
+    first_line = 'HTTP/1.1 {} {}'.format(error_code, error_message)
+    timestamp = 'Date: ' + email.utils.formatdate(usegmt=True)
+    content_header = 'Content-Type: text/plain'
+    body = '{} {}'.format(error_code, error_message)
+    response_list = [first_line, timestamp, content_header, ' ', body, '\r\n']
+
+    recieve = ec.client_socket_function('\r\n'.join(client_request))
+    assert recieve == '\r\n'.join(response_list)
+
+
+class RequestError(Exception):
+    """Exception raised for errors in the request."""
+
+    def __init__(self, value):
+        self.value = value
+
+    def __str__(self):
+        return repr(self.value)
```

Comment:
Why re-define this error class in your tests instead of importing it from the server code module?